components/base/wheeled/sensorbase.go

```diff
@@ -0,0 +1,396 @@
+
+// setPolling determines whether we want the sensor loop to run and stop the base with sensor feedback
+// should be set to false everywhere except when sensor feedback should be polled
+// currently when a orientation reporting sensor is used in Spin.
+func (sb *sensorBase) setPolling(isActive bool) {
+	sb.sensorLoopMu.Lock()
+	defer sb.sensorLoopMu.Unlock()
+	sb.sensorLoopPolling = isActive
+}
+
+// isPolling gets whether the base is actively polling a sensor.
+func (sb *sensorBase) isPolling() bool {
+	sb.sensorLoopMu.Lock()
+	defer sb.sensorLoopMu.Unlock()
+	return sb.sensorLoopPolling
+}
+
+// Spin commands a base to turn about its center at a angular speed and for a specific angle.
+func (sb *sensorBase) Spin(ctx context.Context, angleDeg, degsPerSec float64, extra map[string]interface{}) error {
+	switch {
+	case int(angleDeg) >= 360:
+		sb.setPolling(false)
+		sb.logger.Warn("feedback for spin calls over 360 not supported yet, spinning without sensor")
+		return sb.wBase.Spin(ctx, angleDeg, degsPerSec, nil)
+	default:
+		ctx, done := sb.opMgr.New(ctx)
+		defer done()
+		// check if a sensor context has been started
+		if sb.sensorLoopDone != nil {
+			sb.sensorLoopDone()
+		}
+
+		sb.setPolling(true)
+		// start a sensor context for the sensor loop based on the longstanding base
+		// creator context
+		var sensorCtx context.Context
+		sensorCtx, sb.sensorLoopDone = context.WithCancel(context.Background())
+		if err := sb.stopSpinWithSensor(sensorCtx, angleDeg, degsPerSec); err != nil {
+			return err
+		}
+
+		// // starts a goroutine from within wheeled base's runAll function to run motors in the background
+		if err := sb.startRunningMotors(ctx, angleDeg, degsPerSec); err != nil {
+			return err
+		}
+
+		wb := sb.wBase.(*wheeledBase)
+		motor := wb.allMotors[0]
+		if err := sb.opMgr.WaitTillNotPowered(ctx, 500*time.Millisecond, motor,
+			func(context.Context, map[string]interface{}) error {
+				return nil
+			},
+		); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (sb *sensorBase) startRunningMotors(ctx context.Context, angleDeg, degsPerSec float64) error {
+	if math.Signbit(angleDeg) != math.Signbit(degsPerSec) {
+		degsPerSec *= -1
+	}
+	return sb.wBase.SetVelocity(ctx,
+		r3.Vector{X: 0, Y: 0, Z: 0},
+		r3.Vector{X: 0, Y: 0, Z: degsPerSec}, nil)
+}
+
+func (sb *sensorBase) stopSpinWithSensor(
+	ctx context.Context, angleDeg, degsPerSec float64,
+) error {
+	// imu readings are limited from 0 -> 360
+	startYaw, err := getCurrentYaw(sb.orientation)
+	if err != nil {
+		return err
+	}
+
+	targetYaw, dir, _ := findSpinParams(angleDeg, degsPerSec, startYaw)
+
+	errBound := boundCheckTarget
+
+	// reset error counter for imu reading errors
+	errCounter := 0
+
+	// timeout duration is a multiplier times the expected time to perform a movement
+	spinTimeEst := time.Duration(int(time.Second) * int(math.Abs(angleDeg/degsPerSec)))
+	startTime := time.Now()
+	timeOut := 5 * spinTimeEst
+	if 5*spinTimeEst < 10*time.Second {
```

Comment:
replace 5*spinTimeEst with `timeOut`. This eliminates a possible mistake in the future if someone wants to change the 5 to another value.